management/src/main/java/org/kaazing/gateway/management/monitoring/configuration/MonitorFileWriter.java

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright (c) 2007-2014 Kaazing Corporation. All rights reserved.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.kaazing.gateway.management.monitoring.configuration;
+
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+
+import org.kaazing.gateway.management.monitoring.service.MonitoredService;
+import org.kaazing.gateway.service.MonitoringEntityFactory;
+
+import uk.co.real_logic.agrona.concurrent.UnsafeBuffer;
+
+/**
+ * MonitoringFileWriter interface responsible with writing data to monitoring files
+ *
+ */
+public interface MonitorFileWriter {
+
+    /**
+     * Method computing monitoring file total length
+     * @return
+     */
+    int computeMonitorTotalFileLength();
+
+    /**
+     * Method returning gateway counter labels buffer
+     * @param buffer
+     * @return
+     */
+    UnsafeBuffer createGatewayCounterLabelsBuffer(ByteBuffer buffer);
+
+    /**
+     * Method returning gateway counter values buffer
+     * @param buffer
+     * @return
+     */
+    UnsafeBuffer createGatewayCounterValuesBuffer(ByteBuffer buffer);
+
+    /**
+     * Method creating service counter labels buffer for service index
+     * @param buffer
+     * @param index
+     * @return
+     */
+    UnsafeBuffer createServiceCounterLabelsBuffer(ByteBuffer buffer, int index);
+
+    /**
+     * Method returning service counter values buffer for service index
+     * @param buffer
+     * @param index
+     * @return
+     */
+    UnsafeBuffer createServiceCounterValuesBuffer(ByteBuffer buffer, int index);
+
+    /**
+     * Method returning gateway monitoring entity factory
+     * @param mappedMonitorFile
+     * @param monitoringDir
```

Comment:
The monitoringDir parameter to this method and the next method is only needed so it can be passed through the MMFSeviceWriter instance into the AgronaMonitoringEntityFactory instance that is created for the gateway counters and for each service. And the only way monitoringDir  is used is in AgronaMonitoringEntityFactory.close() which actually deletes the MMF and monitoring directory. This will not work because we must not delete the MMF and monitoring directory until all services have been shut down during Gateway shutdown. I would strongly recommend removing the monitoringDir parameter to this method and the following method and make it the responsibility of the code external to MonitorFileWriter to manage the complete life cycle of the MMF and monitoring directory, in other words, that code should be in charge of deleting those objects as well as creating them. A suitable place to put the logic might be In the stoppedGateway method in ManagementGatewayListener.